DESIGN.rst

```diff
@@ -0,0 +1,156 @@
+
+Blockstore does not:
+
+* Know about courses, libraries, or other "learning contexts"
+* Understand OLX or XBlocks
+* Contain an XBlock runtime
+* Enforce detailed permissions (Instead, the application consuming the Blockstore API, i.e. the LMS, should define its own access control rules and can give out signed links that allow only authorized users to download content files directly from Blockstore)
+* Store learner "state" like what answer was selected or a learner's grade.
+
+Implications
+------------
+
+Because of what's in scope and out of scope for Blockstore as described above, Blockstore itself is generally not used directly by end users. Instead, applications like the LMS or Studio interact with Blockstore and use it much like a database or object storage layer. Those applications are responsble for things like authentication, validation, implementing specific types of content (how is an XBlock represented in Blockstore? how is a content library represented in Blockstore?), etc.
+
+
+--------------------
+Content Libraries v2
+--------------------
+
+So far, the only major feature in Open edX that has been implemented using Blockstore is **Content Libraries v2** (also called "Complex Libraries").
+
+Trying out Content Libraries v2
+-------------------------------
+
+To use or test Content Libraries v2, you need to set up Blockstore, and you'll also want to use the `Library Authoring micro-frontend <frontend-app-library-authoring>`_, which provides an optional user interface for creating content libraries and XBlocks in Blockstore. Follow the instructions in the Library Authoring micro-frontend README for details.
+
+Keys and Identifiers
+--------------------
+
+Every Content Library (v2) is uniquely identified by a `LibraryLocatorV2 <https://github.com/openedx/opaque-keys/blob/5d730556ccdb6e9d7263a94399b9a0897755ac58/opaque_keys/edx/locator.py#L1495>`_ identifier. As a string (e.g. when used in URLs), these identifiers look like ``lib:MITx:reallyhardproblems`` where ``MITx`` is the organization that created the library, and ``reallyhardproblems`` is unique name for that particular content library.
+
+Every XBlock in a Content Library (v2) is uniquely identified by a `LibraryUsageLocatorV2 <https://github.com/openedx/opaque-keys/blob/5d730556ccdb6e9d7263a94399b9a0897755ac58/opaque_keys/edx/locator.py#L1561>`_. As a string (e.g. when used in URLs), these identifiers look like ``lb:MITx:reallyhardproblems:problem:problem1`` where "lb" stands for "library block", the first two parts are the same as the corresponding parts of the library identifier, the third part (``problem``) specifies the type of the XBlock, and the fourth part is a unique name/identifier for that particular XBlock in that library.
+
+Learner Interaction with XBlocks in Content Libraries v2
```

Comment:
Hmm, is there a better place to document the libraries/runtime details or is this the best place?

Reply:
It is already documented in places like https://github.com/openedx/edx-platform/tree/master/openedx/core/djangoapps/xblock#xblock-app-suite-new and the docstrings of files in [`content_libraries`](https://github.com/openedx/edx-platform/tree/master/openedx/core/djangoapps/content_libraries) but I thought it could be a good idea to include a copy of those details here too just because people who don't know the overall architecture may look for them here and the term "blockstore" is often used to refer to both blockstore + this runtime. Let me know if you think it really doesn't belong here or if this should just link back to that though.

Reply:
Got it. Makes sense. 👍🏽